pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java

```diff
@@ -174,7 +176,12 @@ public PhysicalColumnIndexContainer(SegmentDirectory.Reader segmentReader, Colum
       }
 
       if (loadFSTIndex) {
-        _fstIndex = new LuceneFSTIndexReader(segmentReader.getIndexFor(columnName, ColumnIndexType.FST_INDEX));
+        PinotDataBuffer buffer = segmentReader.getIndexFor(columnName, ColumnIndexType.FST_INDEX);
+        if (indexLoadingConfig.getFstIndexType() == FSTType.NATIVE) {
```

Comment:
(major) This is incorrect. We need to create the proper FST index reader based on the header info from the buffer instead of relying on the table config because the table config might have changed since the segment is generated. See how range index is loaded as an example.